File: example/helloworld/go/greeter_client.go
package main

import (
	"fmt"
	pb "mrpc/example/helloworld/go/helloworld"
	"sync"
)

func main() {
	client := pb.NewGreeterClient("127.0.0.1:8080")
	defer client.Close()
	{
		message, err := client.SayHello(&pb.SayHelloRequest{Name: "sync RPC"})
		if err == nil {
			fmt.Println("Greeter 1 received: ", message)
		} else {
			fmt.Println("Greeter 1 failed: ", err)
		}
	}
	{
		key, err := client.AsyncSayHello(&pb.SayHelloRequest{Name: "async RPC"})
		if err == nil {
			message, err := client.Receive(key)
			if err == nil {
				fmt.Println("Greeter 2 received: ", message)
			} else {
				fmt.Println("Greeter 2 receive failed: ", err)
			}
		} else {
			fmt.Println("Greeter 2 send failed: ", err)
		}
	}
	{
		var wg sync.WaitGroup
		wg.Add(1)
		client.CallbackSayHello(&pb.SayHelloRequest{Name: "callback RPC"}, func(message string, err error) {
			defer wg.Done()
			if err == nil {
				fmt.Println("Greeter 3 received: ", message)
			} else {
				fmt.Println("Greeter 3 failed: ", err)
			}
		})
		wg.Wait()
	}
}

File: example/helloworld/python/greeter_client.py
import threading
from helloworld_mrpc import GreeterClient, SayHelloRequest


def main():
    client = GreeterClient("127.0.0.1:8080")

    # 同步调用
    message, err = client.SayHello(SayHelloRequest(name="sync RPC"))
    if err is None:
        print(f"Greeter 1 received: {message}")
    else:
        print(f"Greeter 1 failed: {err}")

    # 异步调用 + Receive
    key, err = client.AsyncSayHello(SayHelloRequest(name="async RPC"))
    if err is None:
        message, err = client.ReceiveSayHello(key)
        if err is None:
            print(f"Greeter 2 received: {message}")
        else:
            print(f"Greeter 2 receive failed: {err}")
    else:
        print(f"Greeter 2 send failed: {err}")

    # 回调调用
    wg = threading.Event()

    def callback(message: str, err: Exception | None):
        if err == None:
            print(f"Greeter 3 received: {message}")
        else:
            print(f"Greeter 3 failed: {err}")
        wg.set()

    client.CallbackSayHello(SayHelloRequest(name="callback RPC"), callback)

    wg.wait()


if __name__ == "__main__":
    main()
